src/dense.jl

```diff
@@ -0,0 +1,243 @@
+  - an object subtyping `EnzymeCore.Mode` (like `EnzymeCore.Forward` or `EnzymeCore.Reverse`) if a specific mode is required
+  - `nothing` to choose the best mode automatically
+
+# Constructors
+
+    AutoEnzyme(; mode=nothing)
+"""
+Base.@kwdef struct AutoEnzyme{M} <: AbstractADType
+    mode::M = nothing
+end
+
+mode(::AutoEnzyme) = ForwardOrReverseMode()  # specialized in the extension
+
+"""
+    AutoFastDifferentiation
+
+Chooses [FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl).
+
+# Constructor
+
+    AutoFastDifferentiation()
+"""
+struct AutoFastDifferentiation <: AbstractADType end
+
+mode(::AutoFastDifferentiation) = SymbolicMode()
+
+"""
+    AutoFiniteDiff{T1,T2,T3}
+
+Chooses [FiniteDiff.jl](https://github.com/JuliaDiff/FiniteDiff.jl).
+
+# Fields
+
+- `fdtype::T1`: finite difference type
+- `fdjtype::T2`: finite difference type for the Jacobian
+- `fdhtype::T3`: finite difference type for the Hessian
+
+# Constructor
+
+    AutoFiniteDiff(; fdtype=Val(:forward), fdjtype=fdtype, fdhtype=Val(:hcentral))
+"""
+Base.@kwdef struct AutoFiniteDiff{T1, T2, T3} <: AbstractADType
+    fdtype::T1 = Val(:forward)
+    fdjtype::T2 = fdtype
+    fdhtype::T3 = Val(:hcentral)
+end
+
+mode(::AutoFiniteDiff) = FiniteDifferencesMode()
+
+"""
+    AutoFiniteDifferences{T}
+
+Chooses [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+
+# Fields
+
+- `fdm::T`: a [`FiniteDifferenceMethod`](https://juliadiff.org/FiniteDifferences.jl/stable/pages/api/#FiniteDifferences.FiniteDifferenceMethod), constructed for instance with `FiniteDifferences.central_fdm`. 
+
+# Constructor
+
+    AutoFiniteDifferences(; fdm)
+"""
+Base.@kwdef struct AutoFiniteDifferences{T} <: AbstractADType
+    fdm::T
+end
+
+mode(::AutoFiniteDifferences) = FiniteDifferencesMode()
+
+"""
+    AutoForwardDiff{chunksize,T}
+
+Chooses [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl).
+
+# Type parameters
+
+- `chunksize`: the preferred [chunk size](https://juliadiff.org/ForwardDiff.jl/stable/user/advanced/#Configuring-Chunk-Size) to evaluate several derivatives at once
+
+# Fields
+
+- `tag::T`: a [custom tag](https://juliadiff.org/ForwardDiff.jl/release-0.10/user/advanced.html#Custom-tags-and-tag-checking-1) to handle nested differentiation calls (usually not necessary)
+
+# Constructors
+
+    AutoForwardDiff(; chunksize=nothing, tag=nothing)
+"""
+struct AutoForwardDiff{chunksize, T} <: AbstractADType
+    tag::T
+end
+
+function AutoForwardDiff(; chunksize = nothing, tag = nothing)
+    AutoForwardDiff{chunksize, typeof(tag)}(tag)
+end
+
+mode(::AutoForwardDiff) = ForwardMode()
+
+"""
+    AutoModelingToolkit
+
+Chooses [ModelingToolkit.jl](https://github.com/SciML/ModelingToolkit.jl).
+
+# Fields
+
+- `obj_sparse::Bool`
+- `cons_sparse::Bool`
+
+# Constructor
+
+    AutoModelingToolkit(; obj_sparse=false, cons_sparse=false)
+"""
+Base.@kwdef struct AutoModelingToolkit <: AbstractADType
+    obj_sparse::Bool = false
```

Comment:
This was based on Optimization.jl's use case, if you plan to have a more general implementation we can change the fields

Reply:
I was wondering about that. Shouldn't this be AutoSymbolics? This "obj" and "cons" doesn't make much sense to me in a generic AD setting. Can you suggest an AutoSymbolics struct to replace it?

Reply:
Yeah `AutoSymbolics` would be more accurate. With `AutoSparse` it should not need any fields

Reply:
Since it's multiple lines I couldn't suggest a change directly

Reply:
doing it now